### PR TITLE
Hook mode solving into Btype backtracking

### PR DESCRIPTION
**Pull request by stedolan**
This patch backtracks changes to mode variables during type errors or Btype.backtrack. In particular, this improves error messages: before this patch, the test case unhelpfully printed:
```
Type int list -> unit is not compatible with type int list -> unit
```
It's best read as separate commits: the first moves all the mode solving stuff to Btype, and the second contains the actual change.

### CODE REVIEW

File: testsuite/tests/typing-local/local.ml [hunk 1]
```diff
@@ -1774,3 +1774,40 @@ Line 1, characters 10-11:
...
+[%%expect{|
+Line 6, characters 46-47:
+6 |   let z : (int list -> unit) -> int -> unit = f
+                                                  ^
+Error: This expression has type
+         (local_ int list -> local_ unit) -> int -> unit
+       but an expression was expected of type
+         (int list -> unit) -> int -> unit
+       Type local_ int list -> local_ unit is not compatible with type
+         int list -> unit
```
Comment (stedolan) [inline testsuite/tests/typing-local/local.ml:1811]: I'm not happy with the `local_ unit` in the return value here, but it's not to do with the change in this patch.

I think `oprint.ml` is printing a `Otyp_arrow (_, Oam_local, _, Oam_unknown, _)` with a local arg and local return, which isn't ideal.

Comment (lpw25) [reply testsuite/tests/typing-local/local.ml:1811]: I think that:
```ocaml
print_out_type_1 mode ppf ty
```
should be:
```ocaml
print_out_type_1 rm ppf ty
```
on line 314 of `oprint.ml`

Comment (stedolan) [reply testsuite/tests/typing-local/local.ml:1811]: Thanks, that makes sense. Patch applied.

**Review by lpw25 — APPROVED** [review]
LGTM

**Pull request merged**